selfie.c

```diff
@@ -5948,34 +5969,74 @@ void emitMalloc() {
   // assuming that page frames are zeroed on boot level zero
   createSymbolTableEntry(LIBRARY_TABLE, (uint64_t*) "zalloc", 0, PROCEDURE, UINT64STAR_T, 0, binaryLength);
 
-  emitLD(REG_A0, REG_SP, 0); // size
+  talloc();
+  talloc();
+
+  emitLD(currentTemporary(), REG_SP, 0); // size
   emitADDI(REG_SP, REG_SP, REGISTERSIZE);
 
-  emitADDI(REG_A7, REG_ZR, SYSCALL_MALLOC);
+  // round up size to double word alignment
+  emitADDI(currentTemporary(), currentTemporary(), SIZEOFUINT64 - 1);
+  emitADDI(nextTemporary(), REG_ZR, SIZEOFUINT64);
+  emitREMU(previousTemporary(), currentTemporary(), nextTemporary());
+  emitSUB(currentTemporary(), currentTemporary(), previousTemporary());
+
+  // call brk syscall to set the current bump pointer
+  emitADD(REG_A0, currentTemporary(), REG_S1);
+  emitADDI(REG_A7, REG_ZR, SYSCALL_BRK);
 
   emitECALL();
 
-  emitJALR(REG_ZR, REG_RA, 0);
+  // return 0 iff memory allocation failed
+  emitBEQ(REG_A0, REG_S1, 2 * INSTRUCTIONSIZE);
+  emitBEQ(REG_ZR, REG_ZR, 4 * INSTRUCTIONSIZE);
+  emitBEQ(REG_ZR, currentTemporary(), 3 * INSTRUCTIONSIZE);
+
+  // no memory was allocated
+  emitADDI(REG_A0, REG_ZR, 0);
+  emitBEQ(REG_ZR, REG_ZR, 4 * INSTRUCTIONSIZE);
+
+  // memory space was allocated
+  emitADDI(currentTemporary(), REG_S1, 0);
+  emitADDI(REG_S1, REG_A0, 0);
+  emitADDI(REG_A0, currentTemporary(), 0);
+
+  tfree(2);
+
+  emitJALR(REG_ZR, REG_RA,0);
 }
 
-void implementMalloc(uint64_t* context) {
+void implementBrk(uint64_t* context) {
   // parameter
+  uint64_t bump;
+
+  uint64_t previousBump;
   uint64_t size;
+  uint64_t valid;
 
-  // local variable
-  uint64_t bump;
+  previousBump = getBumpPointer(context);
```

Comment:
Do this later right before it is required.

Reply:
34ec8fe1ccf452fbad23e7e338d5916324193831